### PR TITLE
fix: revert "chore: Cleanup database sessions (#10427)"

### PR DESCRIPTION
**Pull request by etr2460**
This reverts commit 7645fc85c3d6676a13ae76ca5133f83d8fb54dbe.

### SUMMARY
#10427 broke saving queries in sql lab with the following error:
```
/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py:2505: SAWarning: Usage of the 'Session.add()' operation is not currently supported within the execution stage of the flush process. Results may not be consistent.  Consider using alternative event listeners or connection-level operations instead.
  "event listeners or connection-level operations instead." % method
ERROR:flask_appbuilder.models.sqla.interface:Add record error: This transaction is closed
Traceback (most recent call last):
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2597, in _flush
    flush_context.execute()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 422, in execute
    rec.execute(self)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 624, in execute
    uow,
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 326, in delete_obj
    _organize_states_for_delete(base_mapper, states, uowtransaction)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 463, in _organize_states_for_delete
    base_mapper, uowtransaction, states
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 1602, in _connections_for_states
    connection = uowtransaction.transaction.connection(base_mapper)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 311, in connection
    self._assert_active()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 304, in _assert_active
    raise sa_exc.ResourceClosedError(closed_msg)
sqlalchemy.exc.ResourceClosedError: This transaction is closed
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/flask_appbuilder/models/sqla/interface.py", line 585, in add
    self.session.commit()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/scoping.py", line 162, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 1036, in commit
    self.transaction.commit()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 503, in commit
    self._prepare_impl()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 482, in _prepare_impl
    self.session.flush()
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2496, in flush
    self._flush(objects)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2637, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 81, in __exit__
    compat.raise_(value, with_traceback=traceback)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2637, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 519, in rollback
    self._assert_active(prepared_ok=True, rollback_ok=True)
  File "/Users/erik_ritter/repos/github.com/incubator-superset/venv/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 304, in _assert_active
    raise sa_exc.ResourceClosedError(closed_msg)
sqlalchemy.exc.ResourceClosedError: This transaction is closed
```

This reverts the PR and attempts to fix the merge conflicts on top of it as I couldn't find any specific line at fault here

### TEST PLAN
CI, saving queries work

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

to: @john-bodley @bkyryliuk @dpgaspar @villebro

### CODE REVIEW

File: superset/sql_lab.py [hunk 1]
```diff
@@ -132,7 +132,8 @@ def session_scope(nullpool: bool) -> Iterator[Session]:
         )
     if nullpool:
         engine = sqlalchemy.create_engine(database_uri, poolclass=NullPool)
-        session_class = sessionmaker(bind=engine)
+        session_class = sessionmaker()
```
Comment (john-bodley) [inline superset/sql_lab.py:135]: @etr2460 this was the only thing which I wonder was wrong. In the [docs](https://docs.sqlalchemy.org/en/13/orm/session_api.html#session-and-sessionmaker) where they recommend:

> The class also includes a method configure(), which can be used to specify additional keyword arguments to the factory, which will take effect for subsequent Session objects generated. This is usually used to associate one or more Engine objects with an existing sessionmaker factory before it is first used:
> 
> ```
> # application starts
> Session = sessionmaker()
>
> # ... later
> engine = create_engine('sqlite:///foo.db')
> Session.configure(bind=engine)
> ```

however we're not doing anything between creating and configuring the session.

**Review by john-bodley — COMMENTED** [review]
@etr2460 I went through the PR and couldn't find the part of the code which is responsible for the regression. Note I speculate that there may be something else at play and thus the session logic is somewhat of a red herring.

**Comment by villebro** [comment]
I can't help but wonder if it's the scoped sessions at play.

File: superset/sql_lab.py (hunk 1 above)
Comment (etr2460) [reply superset/sql_lab.py:135]: I don't believe this code is even run when trying to save a query. the session used here is only for running the query i thought?

**Review by john-bodley — APPROVED** [review]
I think reverting makes sense for now.

@villebro et al. maybe we should sync up at some stage to try to think through what the issues could be. Ideally the whole session logic (which is quite convoluted/complex as it involves Superset, FAB, Flask-SQLAlchemy, SQLAlchemy, the DBAPI, and the database client) should probably be handled/administered by Flask-SQLAlchemy as my concern is somewhere we're doing something which is causing issues (possibly related to connections wrongfully being shared) and Flask-SQLAlchemy et al. should prevent this from happening if used correctly.

**Pull request merged**